design/restful-interface

```diff
@@ -0,0 +1,390 @@
+# Vela RESTful API design
+
+Below, I list the Vela backend RESTful API grouped by each resource.
+
+The first version does not use authentication or authorization. We will use go structure to 
+represent request and response body.
+
+## Common properties
+
+### Base URL
+The base URL should be **`vela.oam.dev`** but in our first version, we assumed that it's `localhost`.
+
+### Common Request Headers
+Below are the common request headers. All fields are required.
+
+| Name          | Type   | Values      | Description          |
+|---------------|--------|--------------|----------------------|
+| x-oam-version | String | "2020-08-15" | The REST API version  that user wants to use|
+| x-oam-client-type| String | "CLI", "Dashboard" |  The type of client    |
+| x-oam-request-timestamp | String | yyyy-MM-dd HH:mm:ss.SS | UTC time of the request|
+| x-oam-client-id |  String  |    N/A        |   The unique client id |
+| x-oam-client-request-id|String|  N/A          |   The unique client request id |
+
+### Common Response Headers
+
+ Name          | Type   | Values      | Description          |
+|---------------|--------|--------------|----------------------|
+| x-oam-client-request-id|String|  N/A          |   The unique client request id |
+| x-oam-response-timestamp | String | yyyy-MM-dd HH:mm:ss.SS | UTC time of the response|
+
+
+### Common Error Status
+| Error code                 |  HTTP status code  | User message                                                                       |
+|----------------------------|------------------|------------------------------------------------------------------------------------|
+| Already Exists             | 409              | The specific resource already exists                                               |
+| Not Found                  | 404              | The specific resource not found                                                    |
+| InvalidHeaderValue         | 400              | The value provided for one of the HTTP headers was not in the correct format.      |
+| InvalidQueryParameterValue | 400              | An invalid value was specified for one of the query parameters in the request URI. |
+| InvalidInput               | 400              | One of the request inputs is not valid.                                            |
+| InternalError              | 500              | The server encountered an internal error. Please retry the request.                |
+
+### Common Query parameters
+**updateMode**
+```go
+type updateMode string
+const fullcontent updateMode = "fullContent"
+const cueTemplate updateMode = "cueTemplate"
+```
+
+**qureyMode**
+```go
+type qureyMode string
+const fullcontent qureyMode = "fullContent"
+const cueTemplate qureyMode = "cueTemplate"
+```
+
+## Environment related API
+Environment is a top level resources in the REST API. 
+### Create 
+**URL** : `/api/env`
```

Comment:
I still insist we should be name in create/update api like this:

```
POST /api/envs/<envName>
```

That'll be more RESTful, as we are create one instance of resource

Reply:
agree